Feature nondet array initialization rebase

hannes-steffenhagen-diffblue opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Adds nondet initialisation for statically sized arrays anywhere, and dynamically sized arrays as function arguments when explicitly requested. Dynamically sized arrays can also have an associated size parameter that will be set to the size of the dynamic array.

This is based on a (rebased) version of https://github.com/diffblue/cbmc/pull/3251, new commits start at 0a590c. Code should be reviewable as is, however commits still need some cleanup and there'll need to be some additional documentation before it's mergeable, hence the WIP tag.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Move object factory parameters to util and add java object factory parameters

This moves object_factory_parameterst to util, and creates a new class derived
from it called java_object_factory_parameterst in java_bytecode. This is to
separate out common object factory options in preparation of adding c object
factory parameters.

danpoe committed:
Add optional source location parameter to get_nondet_bool()

danpoe committed:
Use get_nondet_bool() in the c nondet symbol factory

danpoe committed:
Remove c_new_tmp_symbol

danpoe committed:
Add c object factory parameters

danpoe committed:
Parse c object factory options

danpoe committed:
Pass object factory parameters through to object factories

danpoe committed:
Add option --min-null-tree-depth to cbmc

danpoe committed:
Add regression tests for nondet struct init

danpoe committed:
Add recursive nondet struct initialization

danpoe committed:
Adapt jbmc unit tests to use java_object_factory_parameterst

danpoe committed:
Update documentation of max_nondet_tree_depth object factory parameter

hannes-steffenhagen-diffblue commented:
Heads up, tests failures seem to be because we apparently expand `NULL` to `(void*)0` now _before_ generating the assert string. Also, I'll address formatting complaints from cpplint/clang-format when I get around to fixing up the commits

danpoe committed:
Add new cbmc regression tests to the script for deleting failing smt2 tests

danpoe commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -928,6 +929,20 @@ void cbmc_parse_optionst::help()
+    " --max-nondet-tree-depth N    limit size of nondet (e.g. input) object tree;\n" /* NOLINT(*) */ \
+    "                              at level N pointers are set to null\n" /* NOLINT(*) */ \
+    " --min-null-tree-depth N      minimum level at which a pointer can first be\n" /* NOLINT(*) */ \
+    " --pointers-to-treat-as-array <identifier,...>  Comma separated list of identifiers that should be initialized" \
```
Missing newline

danpoe commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -928,6 +929,20 @@ void cbmc_parse_optionst::help()
+    " --pointers-to-treat-as-array <identifier,...>  Comma separated list of identifiers that should be initialized" \
+    "                               as arrays" \
+    "                              NULL in a recursively nondet initialized struct\n" /* NOLINT(*) */ \
+    " --associated-array-sizes <identifier:identifier...>"
```
Missing newline

danpoe commented on regression/cbmc/pointer-to-array-function-parameters-multi-arg-right/test.c:
```diff
@@ -0,0 +1,22 @@
+  {
+    return 0;
+  }
+  // oh no! we should have used prefix_size here
```
Should be removed

hannes-steffenhagen-diffblue replied:
Ah sorry, this was meant to be documentation for what's "wrong" with this test case. Maybe too tongue-in-cheek, I'll reword it

danpoe replied:
I think the wording is fine actually, but it seems the comment is wrong for this file (looks like it's been copied from another file).

hannes-steffenhagen-diffblue replied:
Oh. You're right, sorry didn't see that!

danpoe commented on src/ansi-c/c_object_factory_parameters.h:
```diff
@@ -0,0 +1,49 @@
+
+  void set(const optionst &options) override;
+
+  std::size_t max_dynamic_array_size = 10;
```
10 seems already a bit high for a default value. Can we use 2?

hannes-steffenhagen-diffblue replied:
We can

danpoe commented on src/ansi-c/c_object_factory_parameters.h:
```diff
@@ -0,0 +1,49 @@
+#define CPROVER_ANSI_C_C_OBJECT_FACTORY_PARAMETERS_H
+
+#include <algorithm>
+#include <string>
```
Can we move `<algorithm>` and `<string>` to the cpp file?

danpoe commented on src/ansi-c/c_object_factory_parameters.h:
```diff
@@ -0,0 +1,49 @@
+
+private:
+  std::vector<irep_idt> pointers_to_treat_as_array;
+  std::set<irep_idt> variables_that_hold_array_sizes;
```
Could `pointers_to_treat_as_array` be a set as well?

danpoe commented on src/ansi-c/c_object_factory_parameters.cpp:
```diff
@@ -0,0 +1,103 @@
+    {
+      std::string array_name;
+      std::string size_name;
+      split_string(array_size_pair, ':', array_name, size_name);
```
We should probably catch the `deserialization_exceptiont` here and then throw an `invalid_user_input_exceptiont`.

danpoe commented on src/util/optional_utilities.h:
```diff
@@ -0,0 +1,23 @@
+/*******************************************************************\
```
Nit: these files are typically called *_utils (not *_utilities).

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -127,18 +133,53 @@ exprt symbol_factoryt::allocate_object(
+        // skip, we'll handle this during array initialisation
+        return;
+      }
+    }
```
Array initialization is done here only if `type.id() == ID_pointer` and `expr.id() == ID_symbol`. However, when recursively initializing a struct a member expression is passed to `gen_nondet_init()`. Does this mean that arrays are only nondet initialized when they are function parameters (but not when they are within recursively initialized structs)?

hannes-steffenhagen-diffblue replied:
Right, this is a planned feature, but I've been told to put that on the backburner for now.

For the record, I was thinking of handling that with having a syntax like `struct_name::member_name` for `--pointers-to-treat-as-arrays` and `--associated-array-sizes` (I'd only allow sizes and pointers that are in the same struct for this). That'd presumably have to use struct tags though, which might be a bit annoying for users.

danpoe commented on src/util/string_utils.cpp:
```diff
@@ -109,7 +110,14 @@ void split_string(
+                                     "' to contain two substrings "
+                                     "delimited by " +
+                                     delim + " but has " +
+                                     std::to_string(result.size())};
```
The end of the error message is a bit unclear. Maybe change to something like "... but have \<n\> substrings".

danpoe reviewed (commented):
Looks good so far, a few minor comments

hannes-steffenhagen-diffblue commented:
Also, turns out test failures where also because SMT2 solver is broken. I'll rebase, reorganise and squash some of the commits in here now, so hopefully it'll be review-ready then

hannes-steffenhagen-diffblue committed:
Add nondet initialisation for statically sized arrays

hannes-steffenhagen-diffblue committed:
Make split_string check&throw instead of assert return value

The 2-string version of split_string previously asserted that the split would
result in the right number of substrings. That made it not very useful for
parsing user input, as using it would require parsing the string before passing
it to split_string.

hannes-steffenhagen-diffblue committed:
Add optional_lookup utility function

This provides a more ergonomic interface over map::find

hannes-steffenhagen-diffblue committed:
Add nondet initialisation for pointer array function arguments

Currently restricted to direct pointer arguments (as opposed to i.e.
pointers in structs). Also allows an associated size parameter for a
dynamically sized array to be initialised as well.

The behaviour is controlled through the options

--pointers-to-treat-as-arrays <ident>[,<ident>]*
--associated-array-sizes <ident>:<ident>[,<ident>:<ident>]*
--max-dynamic-array-size <size>

hannes-steffenhagen-diffblue committed:
Document deferred_array_sizes parameter

hannes-steffenhagen-diffblue committed:
Check max dynamic array size parameter
